### PR TITLE
refactor: use primordials

### PR DESCRIPTION
**Pull request by bcoe**
Modules in Node.js use primordial helpers, so that core APIs continue working if prototypes are mutated. By moving to using this same approach it will be easier to drop this module into Node.js once ready.

I believe we will want to do some similar work as part of the resolution to #6, to make the errors we generate follow Node's conventions -- my hope is we can keep this module a direct mirror of the code in `utils/parseargs.js` (in the Node codebase).

### CODE REVIEW

File: primordials.js [hunk 1]
```diff
@@ -0,0 +1,432 @@
+'use strict';
+
+/* eslint-disable node-core/prefer-primordials */
```
Comment (bcoe) [inline primordials.js:3]: This is identical to the `primordials.js` in the Node.js codebase.

File: index.js [hunk 2]
```diff
@@ -1,5 +1,18 @@
 'use strict';
 
+const {
+  ArrayIsArray,
+  ArrayPrototypeConcat,
+  ArrayPrototypeIncludes,
```
Comment (bcoe) [inline index.js:6]: This prevents monkey patched prototypes on types like Array from breaking our core functionality, and is an approach used widely in the Node codebase.

File: primordials.js [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+primordials.AsyncIteratorPrototype =
+  primordials.ReflectGetPrototypeOf(
+    primordials.ReflectGetPrototypeOf(
+      async function* () { }).prototype);
+
+ObjectSetPrototypeOf(primordials, null);
+ObjectFreeze(primordials);
+
+module.exports = primordials;
```
Comment (bcoe) [inline primordials.js:432]: The one change I made from Node's `primordials.js` was defining and exporting `primordials`, I'm guessing in the context of Node it's already available as a global.

File: index.js [hunk 4]
```diff
@@ -53,37 +68,44 @@ const parseArgs = (
...
-    if (arg.startsWith('-')) {
+    if (StringPrototypeStartsWith(arg, '-')) {
       // Everything after a bare '--' is considered a positional argument
       // and is returned verbatim
       if (arg === '--') {
-        result.positionals.push(...argv.slice(++pos));
+        ArrayPrototypePush(
+          result.positionals,
+          ...ArrayPrototypeSlice(argv, ++pos)
+        );
```
Comment (ljharb) [inline index.js:78]: using spread allows hijacking with Array.prototype[Symbol.iterator]
```suggestion
        result.positionals = ArrayPrototypeConcat(
          result.positionals,
          ArrayPrototypeSlice(argv, ++pos)
        );
```

File: package.json [hunk 5]
```diff
@@ -7,7 +7,7 @@
     "coverage": "c8 --check-coverage node test/index.js",
     "test": "c8 node test/index.js",
     "posttest": "eslint index.js",
```
Comment (ljharb) [inline package.json:9]:
```suggestion
    "posttest": "eslint .",
```

File: package.json [hunk 6]
```diff
@@ -7,7 +7,7 @@
...
-    "fix": "eslint index.js --fix"
+    "fix": "eslint *.js --fix"
```
Comment (ljharb) [inline package.json:10]:
```suggestion
    "fix": "npm run posttest -- --fix"
```

File: package.json [hunk 7]
```diff
@@ -25,5 +25,14 @@
     "eslint": "^8.2.0",
     "eslint-plugin-node-core": "iansu/eslint-plugin-node-core",
     "tape": "^5.2.2"
+  },
+  "files": [
+    "index.js",
+    "primordials.js"
+  ],
```
Comment (ljharb) [inline package.json:32]: i strongly prefer not to use the "files" field - it's very dangerous. let's stick with `.npmignore` here.

File: package.json [hunk 8]
```diff
@@ -25,5 +25,14 @@
...
+  "eslint": {
+    "env": {
+      "es6": true
+    }
   }
```
Comment (ljharb) [inline package.json:37]: this should be in `.eslintrc`, and not in package.json at all - both because we already have one, and also because config inline in package.json unavoidably inflates package size.

File: primordials.js [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+  'MathMin',
+  'StringPrototypeConcat',
+  'TypedArrayOf',
+];
+
+function getNewKey(key) {
+  return typeof key === 'symbol' ?
+    `Symbol${key.description[7].toUpperCase()}${key.description.slice(8)}` :
+    `${key[0].toUpperCase()}${key.slice(1)}`;
+}
```
Comment (ljharb) [inline primordials.js:58]: `.description` and `.slice` and `.toUppercase` here all are vulnerable

**Review by ljharb — CHANGES_REQUESTED** [review]
You may want to use https://npmjs.com/call-bind, and `call-bind/callBound`, for something published - node's own patterns only work because node can guarantee first execution for its primordials file, which this package can not.

**Comment by bcoe** [comment]
> You may want to use https://npmjs.com/call-bind, and call-bind/callBound, for something published - node's own patterns only work because node can guarantee first execution for its primordials file, which this package can not.

Ideally I would like to keep `index.js` able to be dropped in to the Node.js codebase, to prevent the two diverging significantly the second we open a PR on Node.js.

Could you submit a patch after this that modifies our `primordials.js` to use `call-bind`?

Now that I think of it, having a separately published shim with the same exports as `primordials.js` in the Node.js codebase would be very handy for future work where a polyfill is developed in parallel.

**Comment by ljharb** [comment]
That’s a good idea. I’ll make such a package today or tomorrow, and then I’ll make a PR here to use it - that way index.js still can be drop-in (which was my thinking with the suggestion anyways; that call-bind would be used in primordials.js), but it’ll actually be robust when used outside of node core.

File: index.js [hunk 10]
```diff
@@ -92,26 +114,29 @@ const parseArgs = (
...
         // shave value as undefined
         result.flags[arg] = true;
         // Append undefined to previous values array for case of
         // multiples option, else add to empty array
-        result.values[arg] = [].concat(
-          options.multiples && options.multiples.includes(arg) &&
+        result.values[arg] = ArrayPrototypeConcat(
+          [],
+          options.multiples && ArrayPrototypeIncludes(options.multiples, arg) &&
             result.values[arg] ?
```
Comment (ljharb) [inline index.js:140]: `a ? a : b` should just be `a || b` (not part of this PR, but always a good fix to make)

File: package.json [hunk 11]
```diff
@@ -6,8 +6,8 @@
   "scripts": {
     "coverage": "c8 --check-coverage node test/index.js",
     "test": "c8 node test/index.js",
-    "posttest": "eslint index.js",
-    "fix": "eslint index.js --fix"
+    "posttest": "eslint *.js",
```
Comment (ljharb) [inline package.json:9]: Why *.js and not just a dot?

Comment (bcoe) [reply package.json:9]: Because it flags a bunch of issues with unused variables in `README.md`, seems like a nuisance to apply the same linting rules to `js` code blocks, since you know they'll be partial code.

File: primordials.js (hunk 9 above)
Comment (bcoe) [reply primordials.js:58]: this will be addressed in follow up work to modify or swap out `primordials.js`.

File: package.json (hunk 11 above)
Comment (ljharb) [reply package.json:9]: I’m confused; how could eslint possibly warn on md files if `--ext=md` isn’t passed?

Either way, we should disable vis eslintignore and/or overrides, not by limiting what gets linted (dot catches nested files, for example, *.js doesn’t)

Comment (shadowspawn) [reply package.json:9]: > I’m confused; how could eslint possibly warn on md files if --ext=md isn’t passed?

Quick comment. Might be layered uses of eslint, like editor extension. I opened a PR prompted by md warnings in my environment! #35

Comment (bcoe) [reply package.json]: I just added README.md to ignores for now. I would have rather just added:

```
<!-- eslint-ignore no-unused-vars !-->
```

Tot he markdown file, which it seems should work based on `markdown/markdown` plugin information ... but none of the incantations I tried fixed it.

File: README.md [hunk 12]
```diff
@@ -119,6 +121,7 @@ const { flags, values, positionals } = parseArgs(argv, options);
 ```
 
 ```js
+const { parseArgs } = require('@pkgjs/parseargs');
```
Comment (bcoe) [inline README.md:124]: This address one of the linting issues, might as well make it so samples are copy/pastable.

**Review by iansu — APPROVED** [review]

**Review by darcyclarke — APPROVED** [review]
👍🏻 LGTM

**Review by ljharb — APPROVED** [review]
LGTM! Sometime this week, I'll make a PR removing `primordials.js` and replacing the require from it with a package that does things properly for "not node core".

**Comment by bcoe** [comment]
> Sometime this week, I'll make a PR removing primordials.js and replacing the require from it with a package that does things properly for "not node core".

@ljharb sounds great, I think this repo will give us a good pattern we can copy for other work.
